### PR TITLE
Handle error in flush_gc_msgs() due to a terminating worker

### LINKED ISSUE #10431: Don't print unhandled errors in tasks if process is terminating
One small addition to https://github.com/JuliaLang/julia/issues/10405 .

If the julia process is terminating due to an `exit()` call or end-of-script, we should ignore any unhandled task failures.

For example, I see this when `exit()` is called from one of my scripts.

```
ERROR (unhandled task failure): ArgumentError: stream is closed or unusable
 in check_open at stream.jl:299
 in uv_write at stream.jl:736
 in flush at stream.jl:790
 in anonymous at multi.jl:179
 in lock at stream.jl:812
 in send_msg_ at ./multi.jl:170
 in remote_do at multi.jl:702
 in flush_gc_msgs at ./multi.jl:163
 in flush_gc_msgs at ./multi.jl:189
 in anonymous at task.jl:478
```

### PR DESCRIPTION
**Pull request by amitmurthy**
Closes https://github.com/JuliaLang/julia/issues/10431

### CODE REVIEW

**Comment by amitmurthy** [comment]
cc: @vtjnash  / @JeffBezanson

**Comment by StefanKarpinski** [comment]
On the contrary, I think we absolutely should print these. If there's an error, it should get handled, even if only to ignore it. Why can't we do that here?

**Comment by amitmurthy** [comment]
Because the errors in this case is due to system trying to do a clean shutdown - handles being closed, finalizers being called, etc. And there is no guarantee in the order in which stuff will be cleaned up.

We don't want to enforce every asynchronous task block to be wrapped in a try-catch and implement throw only if system-not-shutting-down code.

**Comment by StefanKarpinski** [comment]
What about printing unhandled errors before starting the shutdown process?

**Comment by amitmurthy** [comment]
The errors happen due to the shutdown process, not before. In #10431 for example, `flush_gc_msgs` which is a long running task (with the same lifetime as the process), throws up an error when it tries to propagate gc of local `RemoteRef` to remote processes.

In normal circumstances, any error in this task started as
```
    @schedule while true
        wait(any_gc_flag)
        flush_gc_msgs()
    end
``` 
should be treated as a serious bug since it can lead to memory leaks, but when a process is shutting down, the error should be ignored.

Any unhandled errors during regular program lifetime are being printed now in https://github.com/JuliaLang/julia/commit/6343d5018f31ca0daf8af1bc325d569465803f10

**Comment by amitmurthy** [comment]
Just to be clear, I agree with you that tasks must try to handle all expected errors. In the `flush_gc_msgs` example above, any error during normal run should be caught, printed and ignored, while during shutdown it should just be ignored. To make this possible in user code, we should expose a function `proc_status()` that returns either `initializing`, `running` or `shutting-down` so that appropriate action can be taken.

That however, does not take away the intent of this PR, which is to not print errors during shutdown when the system is expected to be in an unpredictable state.

**Comment by jakebolewski** [comment]
Is there a cleaner way to shutdown this Task?  Could a teardown procedure send a sentinel gc message which would cause the task to terminate gracefully instead of forcibly interrupting it?

**Comment by amitmurthy** [comment]
@jakebolewski I don't know the answers to your queries.

I have updated this PR to export `isexiting()` and print any error due to `flush_gc_msgs`

**Comment by ivarne** [comment]
Do you plan to document `isexiting` as well?

**Comment by amitmurthy** [comment]
Yes. After there is consensus on this approach and the name.

**Comment by amitmurthy** [comment]
@StefanKarpinski , could you take a look at the updated code?

File: base/exports.jl [hunk 1]
```diff
@@ -1111,6 +1111,7 @@ export
     atexit,
     clipboard,
     exit,
+    isexiting,
```
Comment (StefanKarpinski) [inline base/exports.jl]: This strikes me as something we shouldn't be exporting. I guess it's an advanced feature, but it feels odd that the user might need to write code that has to know whether it's exiting or not.

**Comment by amitmurthy** [comment]
I am fine with not exporting. But I have come across situations writing finalizers where it would be useful to know if the system is in a proper "running" state or not.

I'll leave it out for now. Can export later if required.

**Comment by amitmurthy** [comment]
During further testing, I saw the same error in `flush_gc_msgs()` when the master was not exiting. At this time, I have changed the PR to only handle the error condition in `flush_gc_msgs()` and not have a generic solution for the process exit case.

**Pull request merged**